stan/math/rev/mat/functor/integrate_1d_tsc.hpp

```diff
@@ -0,0 +1,179 @@
+#ifndef STAN_MATH_PRIM_ARR_FUNCTOR_integrate_1d_tsc_HPP
+#define STAN_MATH_PRIM_ARR_FUNCTOR_integrate_1d_tsc_HPP
+
+#include <stan/math/prim/scal/meta/is_constant_struct.hpp>
+#include <stan/math/prim/mat/fun/value_of.hpp>
+#include <stan/math/rev/scal/fun/value_of.hpp>
+#include <stan/math/rev/scal/fun/to_var.hpp>
+#include <stan/math/rev/mat/fun/to_var.hpp>
+#include <stan/math/rev/arr/fun/to_var.hpp>
+#include <stan/math/prim/scal/err/check_finite.hpp>
+
+#include <stan/math/prim/scal/meta/scalar_seq_view.hpp>
+#include <stan/math/prim/scal/meta/operands_and_partials.hpp>
+#include <stan/math/rev/mat/functor/de_integrator.hpp>
+#include <functional>
+#include <cmath>
+#include <ostream>
+#include <vector>
+
+namespace stan {
+
+  namespace math {
+
+    /**
+     * Return the numeric integral of a function f given its gradient g.
+     *
+     * @tparam T Type of f.
+     * @tparam G Type of g.
+     * @param f a functor with signature
+     * double (double, std::vector<T_param>) or with signature
+     * double (double, T_param) where the first argument is one being
+     * integrated and the second one is either an extra scalar or vector
+     * being passed to f.
+     * @param g a functor with signature
```

Comment:
Why are funtor and gradient being passed separately?  I'd think they'd be much easier to compute together.

Reply:
You mean requesting F to return a vector with its value and gradients all together? Makes a lot of sense and a lot of problems might have common structures between them which could reused to greatly speed up calculations. I'm a little busy this week, so I'll try to finish this PR next weekend.

Reply:
Unfortunatelly I checked that this is not possible, at least not in the current implementation: the function and every single one of its gradients must be integrated separately.